### DISCUSSION TITLE
Use rust-analyzer from path if available?

### DISCUSSION THREAD

**Discussion by Pythoner6** [nix-community/nixvim#1135] [Q&A] [answered]
Is there a way to get the rust-analyzer lsp plugin to use a rust-analyzer from the path if it's available? I have a problem right now I have my machine on nixpkgs stable, and I'm working on a project using https://github.com/ipetkov/crane to use a more recent rust version, and in my dev shell I have it setup to have these newer versions of cargo, rustc, rust-analyzer etc. However in neovim, it still seems to use the version of rust-analyzer in nixpkgs stable, causing errors like this:
```
[ERROR][2024-02-19 11:42:13] .../vim/lsp/rpc.lua:734  "rpc" "/nix/store/fr8j0w5wfrv9fb75f0b0pgdnznr8b563-rust-analyzer-2023-11-13/bin/rust-analyzer"  "stderr"  "[ERROR rust_analyzer::reload] Failed to run proc-macro server from path /nix/store/fjldxvm71jchqi926cjjz5a0h78zb0n6-rust-default-1.76.0/libexec/rust-analyzer-proc-macro-srv, error: Custom { kind: Other, error: \"proc-macro server's api version (3) is newer than rust-analyzer's (2)\" }\n"
```
I'd really like for it to just use rust-analyzer from the path, that way if I work on different projects with different rust toolchains, it can pick up the appropriate one to use for that project automatically.

**Comment by traxys** [comment]
It depends a bit on the way you use rust-analyzer, but if using pure nvim-lspconfig you can set `plugins.lsp.rust-analyzer.package = null;`.

**Reply by Pythoner6** [reply, marked as answer]
Ohh, I see I was missing a bunch of common settings for lsp servers defined in `plugins/lsp/helpers.nix`. Back on 23.11 branch `plugins.lsp.servers.rust-analyzer.package` can't be set to null, but I can set `.installLanguageServer = false` and that combined with making sure my dev shell also had a rust toolchain with `rust-src` enabled did the trick.
